### ISSUE TITLE
FTBFS on Debian/KFreeBSD

### ISSUE THREAD

**Issue by aircrack-ng** [aircrack-ng/aircrack-ng#1074]
**Reported by clopez on 4 Mar 2013 20:32 UTC**

Hi!

I have tested current trunk (aircrack-ng/aircrack-ng@dbb92f289c10cb9e9254a016c6032fa67a20acdb) on Debian/KFreeBSD and it fails to build from source:

```
# make unstable=true sqlite=true
make -C src all
make[Entering directory `/root/test2/aircrack-ng/src'
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude   -c -o aircrack-ng.o aircrack-ng.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude   -c -o crypto.o crypto.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude   -c -o common.o common.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude   -c -o uniqueiv.o uniqueiv.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude   -c -o aircrack-ptw-lib.o aircrack-ptw-lib.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude  aircrack-ng.o crypto.o common.o uniqueiv.o aircrack-ptw-lib.o sha1-sse2.S -o aircrack-ng -lpthread -lssl -lcrypto  -L/usr/local/lib -lsqlite3
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude   -c -o airdecap-ng.o airdecap-ng.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude  airdecap-ng.o crypto.o common.o -o airdecap-ng -lssl -lcrypto 
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude   -c -o packetforge-ng.o packetforge-ng.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude  packetforge-ng.o common.o crypto.o -o packetforge-ng -lssl -lcrypto 
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude   -c -o ivstools.o ivstools.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude  ivstools.o common.o crypto.o uniqueiv.o -o ivstools -lssl -lcrypto 
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude   -c -o kstats.o kstats.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude  kstats.o -o kstats 
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude   -c -o makeivs-ng.o makeivs-ng.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude  makeivs-ng.o common.o uniqueiv.o -o makeivs-ng 
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude   -c -o airdecloak-ng.o airdecloak-ng.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude   -c -o osdep/radiotap/radiotap-parser.o osdep/radiotap/radiotap-parser.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude  airdecloak-ng.o common.o osdep/radiotap/radiotap-parser.o -o airdecloak-ng 
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude   -c -o wpaclean.o wpaclean.c
make -C osdep
make[2](1]:): Entering directory `/root/test2/aircrack-ng/src/osdep'
Building for GNU-kFreeBSD
make[Entering directory `/root/test2/aircrack-ng/src/osdep'
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -fPIC -I..    -c -o osdep.o osdep.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -fPIC -I..    -c -o network.o network.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -fPIC -I..    -c -o file.o file.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -fPIC -I..    -c -o freebsd.o freebsd.c
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -fPIC -I..    -c -o freebsd_tap.o freebsd_tap.c
ar cru libosdep.a  osdep.o network.o file.o freebsd.o freebsd_tap.o
ranlib libosdep.a 
touch .os.GNU-kFreeBSD
make[3](3]:): Leaving directory `/root/test2/aircrack-ng/src/osdep'
make[Leaving directory `/root/test2/aircrack-ng/src/osdep'
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude  wpaclean.o -o wpaclean -Losdep -losdep 
osdep/libosdep.a(file.o): In function `file_read':
/root/test2/aircrack-ng/src/osdep/file.c:86: undefined reference to `ieee80211_radiotap_iterator_init'
/root/test2/aircrack-ng/src/osdep/file.c:89: undefined reference to `ieee80211_radiotap_iterator_next'
collect2: error: ld returned 1 exit status
make[1](2]:): *** [Error 1
make[1](wpaclean]): Leaving directory `/root/test2/aircrack-ng/src'
make: *** [Error 2
```

After some tests I fixed it by adding osdep/radiotap/radiotap-parser.o to the list of objects to be linked for wpaclean. But then another airodump-ng binaries failed with the same errors. So I added this object to all the binaries that failed to build.

After this, all binaries built correctly.

However, besside-ng gave a warning

```
gcc -g -W -Wall -O3 -D_FILE_OFFSET_BITS=64 -D_REVISION=0  -I/usr/local/include -DHAVE_SQLITE -Wno-unused-but-set-variable -Wno-array-bounds -Iinclude -DUSE_GCRYPT   -c -o besside-ng.o besside-ng.c
besside-ng.c: In function cracker_kill:
besside-ng.c:277:4: warning: implicit declaration of function close [-Wimplicit-function-declaration](all])
besside-ng.c: In function open_pcap:
besside-ng.c:765:2: warning: implicit declaration of function write [In function wpa_upload:
besside-ng.c:879:2: warning: implicit declaration of function lseek [-Wimplicit-function-declaration](-Wimplicit-function-declaration]
besside-ng.c:)
besside-ng.c:891:3: warning: implicit declaration of function read [In function wpa_crack:
besside-ng.c:950:2: warning: implicit declaration of function fsync [-Wimplicit-function-declaration](-Wimplicit-function-declaration]
besside-ng.c:)
besside-ng.c: In function cracker_start:
besside-ng.c:2325:2: warning: implicit declaration of function pipe [warning: implicit declaration of function fork [-Wimplicit-function-declaration](-Wimplicit-function-declaration]
besside-ng.c:2328:2:)
```

I fixed it by adding a missing include.

Now all builds correctly on Debian/KFreeBSD and the tests pass without problem.

I did some QA and tested to build aircrack-ng with this patch on Linux and did some captures and decrypted them with airdecap-ng and everything is working as expected.

**Comment by aircrack-ng** [comment]
**Attachment by clopez on 4 Mar 2013 20:33 UTC**

Uploaded file [fix-FTBFS-on-DebianKFreeBSD.patch](https://github.com/aircrack-ng/trac-attachments/blob/master/1074/20130304-203308_000-fix-FTBFS-on-DebianKFreeBSD.patch)

**Comment by aircrack-ng** [comment]
**Comment by clopez on 4 Mar 2013 20:32 UTC**

ACK?

**Comment by aircrack-ng** [comment]
**Comment by clopez on 4 Mar 2013 20:32 UTC**

I have just tested building current trunk (aircrack-ng/aircrack-ng@83517459ae2effb862f3be5613974ba8a33065df) on a machine running Debian/HURD and the same error happens. The proposed patch fixes this as well, after applying it, the build and tests run without problems.

**Comment by aircrack-ng** [comment]
**Comment by misterx on 4 Mar 2013 20:32 UTC**

ACKed.

**Comment by aircrack-ng** [comment]
**Comment by clopez on 4 Mar 2013 20:32 UTC**

Fixed on aircrack-ng/aircrack-ng@7a9e87c99a5113d651c94b4aade890420fcc4e9c
